src/core/net/ip6.cpp

```diff
@@ -83,6 +83,9 @@ Ip6::Ip6(Instance &aInstance)
     , mTcp(aInstance)
 #endif
 {
+#if OPENTHREAD_CONFIG_IP6_BR_COUNTERS_ENABLE
+    ResetBorderRoutingCounters();
```

Comment:
I wonder adding a constructor to `otBorderRoutingCounters` class. may have better readability and performance.

Reply:
I think we shouldn't do that because  `otBorderRoutingCounters` is a part of C (instead of C++) API.

Reply:
All public OT types/APIs are defined in C. The `otBorderRoutingCounters` is  a C struct so we cannot add constructor to it as is.

Reply:
We can define a mirror C++ type (we do it often for other C types, e.g. `ot::Mac::Address` for `otMacAddress`), and then can add constructor to the new OT core C++ type (we can then also maybe inherit from `Clearable<>` to get `Clear()` method and use it).

I think the case here, I am not sure if adding new type so to add a constructor would make overall code simpler (if we need to add more methods to the new type then maybe it make sense to define it). As example, we don't do this with `otMleCounters`.

Reply:
BTW, we used to have many usage of memset in early stage of Matter code, but almost all of them have been replaced by RAII style code.

Modern C++ feature can greatly increase code readibility and also improve robustness.

Reply:
Anyway, to move this PR forward, we may need to replace the `memset` with manually resetting all field of the structure.

Reply:
It is also fine to leverage it into another PR, at you will.

Reply:
> Both C and C++ allow following member field being put inside the padding of the last member, if it's size fits, for example, squeeze a boolean into a padding with size 1. 

I  expected/assumed that this would not be allowed for PoD (or standard-layout) types (which most our types are, specially the C structs) as they are expected to be `memcpy` able. Did some investigation and it does look like "padding reuse" is allowed even with PoD member types (they would still be technically `memcpy`-able)

We do use `Clearable<>` (which does use `memset(0)`) in a lot of types on OT core. I guess this may be fine for types that are declared as "packed", but then not safe for other types in general. 

It may be good to update this later to make it safer but I am afraid may require a lot of changes and also I expect it will  increase code size (as we may now need to explicitly set all member variable to zero to implement `Clear()`.

Reply:
Agree, we can check it later.

Reply:
I have been reading and looking more into the C++ standard.
- We have the notion of ["trivially copyable" types](https://en.cppreference.com/w/cpp/language/classes#Trivially_copyable_class)
  - This basically allows bytes from the object of such types to be copied into another object (to copy it).
  - And we can get trivial copy/assignment operator (implicitly defined) which [spec mentions would do `memmove`](https://en.cppreference.com/w/cpp/language/copy_assignment).
  
I believe our type `otBorderRoutingCounters` meets all requirements so it is "trivially copyable" and we do get an implicit trivial copy assignment for it.
- If the compiler tries to do "padding reuse" optimization for this type and inserts a new variable into `mBorderRoutingCounters` variable padding, then it can break the trivial copy assignment use for this variable:
   - We expose it as reference in `GetBorderRoutingCounters()`
   - And it is valid to have a statement as `GetBorderRoutingCounters() = someOtherCountersInstance;` (use assign operator) somewhere in the code (maybe in another compilation unit)
   - It would be unsafe for complier to do "padding reuse" for such a type.
   - So while it is not explicitly disallowed, I think it can not be practically used.